invoices/sql_migration.go

```diff
@@ -190,3 +193,99 @@ func MigrateSingleInvoice(ctx context.Context, tx SQLInvoiceQueries,
 
 	return nil
 }
+
+// MigrateInvoices migrates all invoices from the old database to the new SQL
+// schema. The migration is done in a single transaction to ensure that all
+// invoices are migrated or none at all.
+func MigrateInvoices(ctx context.Context, db InvoiceDB,
+	sqlStore *SQLStore, netParams *chaincfg.Params, batchSize int) error {
+
+	offset := uint64(0)
+	var ops SQLInvoiceQueriesTxOptions
+	return sqlStore.db.ExecTx(ctx, &ops, func(tx SQLInvoiceQueries) error {
+		for {
+			query := InvoiceQuery{
+				IndexOffset:    offset,
+				NumMaxInvoices: uint64(batchSize),
+			}
+
+			queryResult, err := db.QueryInvoices(ctx, query)
+			if err != nil {
+				return fmt.Errorf("unable to query invoices: "+
+					"%v", err)
+			}
+
+			if len(queryResult.Invoices) == 0 {
+				log.Infof("All invoices migrated")
+
+				return nil
+			}
+
+			err = migrateInvoices(
+				ctx, tx, sqlStore, queryResult.Invoices,
+				netParams,
+			)
+			if err != nil {
+				return err
+			}
+
+			offset = queryResult.LastIndexOffset
+		}
+	}, func() {})
+}
+
+func migrateInvoices(ctx context.Context, tx SQLInvoiceQueries,
+	sqlStore *SQLStore, invoices []Invoice,
+	netParams *chaincfg.Params) error {
+
+	for i, invoice := range invoices {
+		var paymentHash lntypes.Hash
+		if invoice.Terms.PaymentPreimage != nil {
+			paymentHash = invoice.Terms.PaymentPreimage.Hash()
+		} else {
```

Comment:
This doesn't correctly handle the case of an invoice generated by receiving a spontaneous AMP payment, which should have an empty payment hash IIUC. To test this, run the migration on Bob's DB resulting from the itest `sendpayment_amp_invoice/external_payaddr`.

Reply:
Nice catch, thank you for pointing this out! Fix is coming soon as well as the rest of the missing parts. Migrating spontaneous AMP invoices required building an index at migration time so we're able to map add indexes to payment hashes.

Reply:
looks like spontaneous amp payments still have a valid payment hash tho  or ?

Reply:
For such invoices we only store the hash in the `paymenthashes` bucket, so we needed to first create a mapping that'd let us query the hash for individual invoices during migration.

Reply:
Ahh got you, so this is not only the cases for spontanous AMP invoices, but also for normal AMP invoices, where the hash is just a random hash, and I think we never learn the preimage as well.